src/main/kotlin/cacophony/controlflow/CFGNode.kt

```diff
@@ -0,0 +1,152 @@
+package cacophony.controlflow
+
+class CFGLabel
+
+sealed class CFGVertex(
+    val tree: CFGNode,
+) {
+    abstract fun dependents(): List<CFGLabel>
+
+    sealed class Conditional(
+        tree: CFGNode.LogicalOperator,
+        val trueDestination: CFGLabel,
+        val falseDestination: CFGLabel,
+    ) : CFGVertex(tree) {
+        override fun dependents() = listOf(trueDestination, falseDestination)
+    }
+
+    sealed class Jump(
+        tree: CFGNode.Unconditional,
+        val destination: CFGLabel,
+    ) : CFGVertex(tree) {
+        override fun dependents() = listOf(destination)
+    }
+
+    sealed class Final(
+        tree: CFGNode.Unconditional,
+    ) : CFGVertex(tree) {
+        override fun dependents() = emptyList<CFGLabel>()
+    }
+}
+
+sealed interface CFGNode {
+    sealed interface Unconditional : CFGNode
+
+    sealed interface Leaf : CFGNode
+
+    sealed class Return :
+        Unconditional,
+        Leaf
+
+    sealed class Call :
+        Unconditional,
+        Leaf
+
+    sealed class Assignment(
+        val regvar: String, // TODO: make Register
+        val value: CFGNode,
+    ) : Unconditional
+
+    sealed class VariableUse(
+        val regvar: String, // TODO: make Register
+    ) : Unconditional,
+        Leaf
+
+    sealed class MemoryAccess(
+        val destination: CFGNode,
+    ) : Unconditional
+
+    sealed class Constant(
+        val value: Int,
+    ) : Unconditional,
+        Leaf
+
+    sealed class Sequence(
+        vararg val nodes: CFGNode,
```

Comment:
I think we should avoid varargs in primary constructors and simply use lists. Maybe add it as a secondary contructor.